src/Generating/FinishGen.cpp

```diff
@@ -180,6 +180,187 @@ void cFinishGenNetherClumpFoliage::TryPlaceClump(cChunkDesc & a_ChunkDesc, int a
+	if (PossibleBlocks.empty())
+	{
+		// No need to go any further. This biome can't generate any blocks.
+		return;
+	}
+
+	int NumClumps = info.m_MaxNumClumpsPerChunk - info.m_MinNumClumpsPerChunk;
+	if (NumClumps == 0)
+	{
+		NumClumps = 1;
+	}
+
+	NumClumps = NoiseVal % NumClumps + info.m_MinNumClumpsPerChunk;
+	int seq = 1;
+	for (int i = 0; i < NumClumps; i++)
+	{
+		int Val1 = m_Noise.IntNoise2DInt(ChunkX * ChunkZ * seq, ChunkZ + ChunkX / seq);
+		int Val2 = m_Noise.IntNoise2DInt(ChunkZ * ChunkX / seq, ChunkZ - ChunkX * seq);
+
+		int PosX = Val1 % (cChunkDef::Width - RANGE_FROM_CENTER * 2) + RANGE_FROM_CENTER;
+		int PosZ = Val2 % (cChunkDef::Width - RANGE_FROM_CENTER * 2) + RANGE_FROM_CENTER;
+
+		int BlockVal = m_Noise.IntNoise2DInt(Val1, Val2);
+
+
+		int TotalWeight = 0;
+		for (auto Block : PossibleBlocks)
+		{
+			TotalWeight += Block.m_Weight;
+		}
+
+		int Weight = BlockVal % TotalWeight;
+		for (auto Block : PossibleBlocks)
+		{
+			Weight -= Block.m_Weight;
+			if (Weight < 0)
+			{
+				TryPlaceFoliageClump(a_ChunkDesc, PosX, PosZ, Block.m_BlockType, Block.m_BlockMeta, Block.m_BlockType == E_BLOCK_BIG_FLOWER);
+				break;
+			}
+		}
+
+		seq++;
+	}
+}
+
+
+
+
+
+void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int a_CenterX, int a_CenterZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, bool a_IsDoubleTall)
+{
+	int ChunkX = a_ChunkDesc.GetChunkX();
+	int ChunkZ = a_ChunkDesc.GetChunkZ();
+
+	int NumBlocks = m_Noise.IntNoise2DInt(a_CenterX + ChunkX * 16, a_CenterZ + ChunkZ * 16) % (MAX_NUM_FOLIAGE - MIN_NUM_FOLIAGE) + MIN_NUM_FOLIAGE + 1;
+	for (int i = 1; i < NumBlocks; i++)
+	{
+		int x = a_CenterX + ((m_Noise.IntNoise2DInt(i * ChunkX * a_CenterX, i * ChunkZ) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
+		int z = a_CenterZ + ((m_Noise.IntNoise2DInt(i * ChunkZ, i * ChunkX * a_CenterZ) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
+		int Top = a_ChunkDesc.GetHeight(x, z);
+
+		if (a_ChunkDesc.GetBlockType(x, Top, z) != E_BLOCK_GRASS)
+		{
+			continue;
+		}
+
+		a_ChunkDesc.SetBlockTypeMeta(x, Top + 1, z, a_BlockType, a_BlockMeta);
+
+		if (a_IsDoubleTall)
+		{
+			a_ChunkDesc.SetBlockTypeMeta(x, Top + 2, z, E_BLOCK_BIG_FLOWER, 8);
+		}
+	}
+
+}
+
+
+
+
+
+std::vector<cFinishGenClumpTopBlock::BiomeInfo> cFinishGenClumpTopBlock::ParseConfigurationString(AString a_String)
+{
+	std::vector<BiomeInfo> BiomeList;
+
+	// Initialize the vector for all biomes.
+	for (int i = 0; i < biMaxVariantBiome; i++)
+	{
+		BiomeList.push_back(BiomeInfo());
+	}
+
+	AStringVector ClumpList = StringSplitAndTrim(a_String, "|");
+	for (auto RawClumpInfo : ClumpList)
+	{
+		AStringVector ClumpInfo = StringSplitAndTrim(RawClumpInfo, ";");
+
+		// Information about a clump is divided in 2 parts. The biomes they can be in and the blocks that can be placed.
+		if (ClumpInfo.size() != 2)
+		{
+			LOGWARNING("Data missing for \"%s\". Please divide biome and blocks with a semi colon", RawClumpInfo);
```

Comment:
All warnings displayed to the console should indicate what part of the configuration is wrong (`OverworldClumpFoliage: ...`)
`divide ... and ...` -> `separate ... from ...`